components/node-labeler/cmd/run.go

```diff
@@ -197,90 +224,187 @@ func (r *PodReconciler) Reconcile(ctx context.Context, req reconcile.Request) (r
-		// nothing to do, the label already exists.
-		return reconcile.Result{}, nil
+		return false, nil
 	}
 
-	err = checkTCPPortIsReachable(ipAddress, port)
+	err := checkTCPPortIsReachable(ipAddress, port)
 	if err != nil {
 		log.WithField("host", ipAddress).WithField("port", port).WithField("pod", pod.Name).WithError(err).Error("checking if TCP port is open")
-		return reconcile.Result{RequeueAfter: defaultRequeueTime}, nil
+		return false, nil
 	}
 
-	if component == registryFacade {
+	if strings.HasPrefix(pod.Name, registryFacade) {
 		err = checkRegistryFacade(ipAddress, port)
 		if err != nil {
 			log.WithError(err).Error("checking registry-facade")
-			return reconcile.Result{RequeueAfter: defaultRequeueTime}, nil
+			return false, nil
 		}
+	}
+
+	return true, nil
+}
 
-		time.Sleep(1 * time.Second)
+type NodeReconciler struct {
+	client.Client
+}
+
+func (r *NodeReconciler) nodeFilter() predicate.Predicate {
+	return predicate.Funcs{
+		CreateFunc: func(e event.CreateEvent) bool {
+			node, ok := e.Object.(*corev1.Node)
+			if !ok {
+				return false
+			}
+			return isWorkspaceNode(*node)
+		},
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			return false
+		},
+		DeleteFunc: func(e event.DeleteEvent) bool {
+			return false
+		},
+	}
+}
+
+func (r *NodeReconciler) reconcileAll(ctx context.Context) error {
+	log.Info("start reconciling all nodes")
+
+	var nodes corev1.NodeList
+	if err := r.List(ctx, &nodes); err != nil {
+		return fmt.Errorf("failed to list nodes: %w", err)
 	}
 
-	err = updateLabel(labelToUpdate, true, nodeName, r)
-	if err != nil {
-		log.WithError(err).Error("updating node label")
-		return reconcile.Result{}, fmt.Errorf("trying to add the label: %v", err)
+	for _, node := range nodes.Items {
+		if node.Labels == nil {
+			continue
+		}
+		if !isWorkspaceNode(node) {
+			continue
+		}
+
+		err := updateNodeLabel(node.Name, r.Client)
+		if err != nil {
+			log.WithError(err).WithField("node", node.Name).Error("failed to initialize labels on node")
+		}
+		r.Reconcile(ctx, reconcile.Request{NamespacedName: types.NamespacedName{Name: node.Name}})
 	}
 
-	readyIn := time.Since(pod.Status.StartTime.Time)
-	NodeLabelerTimeHistVec.WithLabelValues(component).Observe(readyIn.Seconds())
-	NodeLabelerCounterVec.WithLabelValues(component).Inc()
+	log.Info("finished reconciling all nodes")
+	return nil
+}
 
+func (r *NodeReconciler) Reconcile(ctx context.Context, req reconcile.Request) (reconcile.Result, error) {
+	var node corev1.Node
+	err := r.Get(ctx, req.NamespacedName, &node)
+	if err != nil {
+		if !errors.IsNotFound(err) {
+			log.WithError(err).Error("unable to fetch node")
+		}
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+	var podList corev1.PodList
+	err = r.List(ctx, &podList, client.MatchingFields{
+		"spec.nodeName": node.Name,
+	})
+	if err != nil {
+		return reconcile.Result{}, fmt.Errorf("listing pods: %w", err)
```

Comment:
```suggestion
		return reconcile.Result{}, fmt.Errorf("cannot list pods: %w", err)
```

Reply:
Am I missing an idiomatic thing? I ask because for an error, it seems to use a mesage that indicates success (hence changing to cannots)

Reply:
Let me know, I'm not sure if you were following existing pattern, or I am missing a practice @iQQBot .